### PR TITLE
Current Development of Alice

### PR DESCRIPTION
**Pull request by t532**
Coding is done as well as further testing.

- ~~HTTP API~~
- ~~Token validation~~
- [x] Judging
- [x] Updating and compiling Submission
- [x] Updating and compiling SpecialJudge
- [x] Apply JS API
- [x] Add support for Partially Correct (PC)
- [x] Further Testing
- ~~Upload testcases through [Tus Protocol](https://tus.io)~~
- ~~Implement HTTPS~~

### CODE REVIEW

File: src/router.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            time?: string
+            memory?: string
+        }
+    }>()
+    _router.delete('/delete_testcase', async ctx => {
+        deleteTestcase(ctx.request.query['testcase_id'])
+    })
+    // BUG: Tus server cannot be put in koa middlewares
+    _router.all('/update_testcase_infile(.*)', async ctx => {
+        updateTestcase(ctx.request.header['testcase-id'], 'in', ctx.req, ctx.res)
```
Comment (t532) [inline src/router.ts]: The `Testcase-Id` HTTP header should be `X-Testcase-Id`.

File: src/router.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            submission_id?: string
+            language?: string
+            time?: string
+            memory?: string
+        }
+    }>()
+    _router.delete('/delete_testcase', async ctx => {
+        deleteTestcase(ctx.request.query['testcase_id'])
+    })
+    // BUG: Tus server cannot be put in koa middlewares
```
Comment (t532) [inline src/router.ts]: Bug here

Comment (t532) [reply src/router.ts]: Wontfix - File Uploading will be deleted in Alice and be moved to WhoJudge API

File: src/router.ts (hunk 1 above)
Comment (t532) [reply src/router.ts]: Wontfix.

**Comment by StephaneZhang** [comment]
That's ok.

**Pull request merged**